modules/videoio/src/cap_libcamera.cpp

```diff
@@ -0,0 +1,1001 @@
+            if (requests_.back()->addBuffer(stream, buffer) < 0)
+                CV_Error(cv::Error::StsAssert, "failed to add buffer to request");
+        }
+    }
+}
+
+void LibcameraCapture::requestComplete(Request *request)
+{
+    if (request->status() == Request::RequestCancelled)
+        return;
+
+    CompletedRequest *r = new CompletedRequest(sequence_++, request);
+    CompletedRequestPtr payload(r, [this](CompletedRequest *cr)
+                                { this->queueRequest(cr); });
+    {
+        std::lock_guard<std::mutex> lock(completed_requests_mutex_);
+        completed_requests_.insert(r);
+    }
+
+    // We calculate the instantaneous framerate in case anyone wants it.
+    uint64_t timestamp = payload->buffers.begin()->second->metadata().timestamp;
+    if (last_timestamp_ == 0 || last_timestamp_ == timestamp)
+        payload->framerate = 0;
+    else
+        payload->framerate = 1e9 / (timestamp - last_timestamp_);
+    last_timestamp_ = timestamp;
+
+    msg_queue_.Post(Msg(MsgType::RequestComplete, std::move(payload)));
+}
+
+void LibcameraCapture::configureDenoise(const std::string &denoise_mode)
+{
+    using namespace libcamera::controls::draft;
+
+    static const std::map<std::string, NoiseReductionModeEnum> denoise_table = {
+        {"off", NoiseReductionModeOff},
+        {"cdn_off", NoiseReductionModeMinimal},
+        {"cdn_fast", NoiseReductionModeFast},
+        {"cdn_hq", NoiseReductionModeHighQuality}};
+    NoiseReductionModeEnum denoise;
+
+    auto const mode = denoise_table.find(denoise_mode);
+    if (mode == denoise_table.end())
+        CV_Error(cv::Error::StsAssert, "Invalid denoise mode " + denoise_mode);
+    denoise = mode->second;
+
+    controls_.set(NoiseReductionMode, denoise);
+}
+
+LibcameraCapture::LibcameraCapture() : LibcameraCapture(-1) {}
+
+LibcameraCapture::LibcameraCapture(int camera_index)
+{
+    options_ = std::unique_ptr<Options>(new Options());
+    controls_ = controls::controls;
+    if (!options_)
+        options_ = std::make_unique<Options>();
+    controls_.clear();
+    options = static_cast<Options *>(GetOptions());
+    still_flags = FLAG_STILL_NONE;
+    options->photo_width = 2560;
+    options->photo_height = 1440;
+    options->video_width = 640;
+    options->video_height = 480;
+    options->framerate = 30;
+    options->denoise = "auto";
+    options->timeout = 1000;
+    options->setMetering(Metering_Modes::METERING_MATRIX);
+    options->setExposureMode(Exposure_Modes::EXPOSURE_NORMAL);
+    options->setWhiteBalance(WhiteBalance_Modes::WB_AUTO);
+    options->contrast = 1.0f;
+    options->saturation = 1.0f;
+    options->camera = camera_index;
+    // still_flags |= FLAG_STILL_RGB;
+    still_flags |= FLAG_STILL_BGR;
+    needsReconfigure.store(false, std::memory_order_release);
+    camerastarted = false;
+    current_request_ = nullptr;
+    // Automatically open the camera
+    if (camera_index >= 0) {
+        open(camera_index);
+    }
```

Comment:
It silently ignores incorrect input. I propose to remove check here and add CV_Assert at the very beginning. No much sense to fill options, if camera index is not correct.